### DISCUSSION TITLE
What does the reader of arrays and hashes return in scalar context?

### DISCUSSION THREAD

**Discussion by yuki-kimoto** [Ovid/Cor#80] [Q&A] [answered]
What does the reader of arrays and hashes return in scalar context?

6.2.3.2 :reader(optional_identifier)

https://github.com/Ovid/Cor/blob/master/rfc/attributes.md#6232-readeroptional_identifier

```
class MyClass {
  field @array : reader { (1, 2, 3) };
  field @hash : reader { (x => 1, y => 2) };
}

my $object = MyClass->new;

my $ret_array = $object->array;

my $ret_hash = $object->hash;
```

1. the first element of the array or the hash. In this case <code>$ret_array</code> is <code>1</code>. <code>$ret_hash</code> is <code>"x"</code>. 
2. The count of the elements of the array or the hash. In this case <code>$ret_array</code> is <code>3</code>. <code>$ret_hash</code> is <code>4</code>.

**Comment by HaraldJoerg** [comment]
This will behave as Perl does if you use an array or hash in scalar context.  A reader in scalar context returns the number of elements for an array field, and the number of _keys_ (since Perl 5.26) for a hash field.  So, `$ret_array` is 3 and `$ret_hash` is 2.

**Comment by Ovid** [comment, marked as answer]
At the present time, we don't plan to put attributes on fields which are not scalars because the behavior isn't clear. This would be a feature for after the MVP.

**Reply by yuki-kimoto** [reply]
Dose MVP contains the reader of the array and the hash?

**Reply by leonerd** [reply]
Readers on array/hash fields are already in `Object::Pad`. I'll put them in `feature-class` if the implementation of them is small and easy. I expect it will be small and easy. So quite likely.

**Reply by yuki-kimoto** [reply]
Thanks. OK, I will read Readers on array/hash fields in <code>Object::Pad</code> today.

**Reply by yuki-kimoto** [reply]
I red them now.

```
field @items :reader;
 
foreach my $item ( $obj->items ) { ... }   # iterates the list of items
 
my $count = $obj->items;                   # yields count of items
```

I assume this behavior in <code>use feature 'class'</code>.

**Reply by yuki-kimoto** [reply]
In my personal opinion, the word "spec" doesn't means the minimal implementation.

The people who implement the spec implements what the spec writes.

**Comment by leonerd** [comment]
@yuki-kimoto You do seem to be asking a lot of questions lately that could be quite easily answered by reading the spec / current implementation / inspiration from Object::Pad. Engagement is good, but perhaps you could do some of your own research first before raising all these questions?

**Reply by yuki-kimoto** [reply]
> You do seem to be asking a lot of questions lately that could be quite easily answered by reading the spec / current implementation / inspiration from Object::Pad. Engagement is good, but perhaps you could do some of your own research first before raising all these questions?

It is possible to do it. 

However, @haarg said 

```
The class feature that LeoNerd is working on is intended to be the implementation of the Corinna spec.
```

https://github.com/Ovid/Cor/discussions/66

I can study Object::Pad, however, how do I know the difference between Corrina and Object::Pad?
